🐛 BUG: "can't use ".." to break out of starting directory" when using Service Bindings in monorepo

### Which Cloudflare product(s) does this pertain to?

Miniflare

### What version(s) of the tool(s) are you using?

3.20231218.1 [miniflare]

### What version of Node are you using?

20

### What operating system are you using?

Mac

### Describe the Bug

I'm attempting to use Miniflare to run one app (app-2) with a second app (app-1) as a Service Binding. The docs are a little thin here, so I'm not sure if I'm defining the workers in the correct way.

```js
return new Miniflare({
  port: 4000,

  workers: [
    {
      name: 'musicfetch-api',
      scriptPath: 'dist/index.js',
      modules: true,

      serviceBindings: {
        APP_2: 'app-2',
      },
    },
    {
      name: 'app-2',
      scriptPath: resolvePath(__dirname, '../app-2/dist/index.js'),
      modules: true,

      durableObjects: {
        MyDurableObject: 'MyClass',
      },
    },
  ],
});
```

In doing this I'm getting the error:

```
workerd/util/symbolizer.c++:98: warning: Not symbolizing stack traces because $LLVM_SYMBOLIZER is not set. To symbolize stack traces, set $LLVM_SYMBOLIZER to the location of the llvm-symbolizer binary. When running tests under bazel, use `--test_env=LLVM_SYMBOLIZER=<path>`.
workerd/jsg/util.c++:281: error: e = kj/filesystem.c++:315: failed: expected parts.size() > 0 [0 > 0]; can't use ".." to break out of starting directory
stack: 1028e4c07 1028e3963 1028e383f 1008c776f 100b7ece3 100da687f 100db41ef 100db4083 100877e67 10087ddc7 1008815c3 10087ff13 100871b03 100871733 100871653 10087163b 1028fb9ff 1028fbd47 1028fa4a7 1028fa277 1008641af 182aed0df; sentryErrorContext = jsgInternalError
service core:user:rate-limiter: Uncaught Error: internal error
```

Notably: `can't use ".." to break out of starting directory`

I've tried various other approaches (eg. importing a Miniflare instance from `app-2` and calling it programatcally), but it results in the same error. It feels like I'm battling against the framework here for what I imagine is a common project structure.

- Is there a way to define the Miniflare CWD? If so I could have it execute in the parent directory.
- For now I will probably hack around this, but just wanted to leave an issue here to confirm there wasn't a better way.

Cheers 🍺 


### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

_No response_

It doesn't work when setting `scriptPath: scriptPath`. Yes, the script contains a dynamic import to `cloudflare:sockets`. It seems like the script file needs to be in the same directory tree as the script that sets up Miniflare. (Our script is in `../worker/dist/index.mjs`.)
 
This is the error (same regardless of whether we're adding `script: await readFile(scriptPath)`):

```
capnp compile log.capnp -oc:capnp
*** Uncaught exception ***
kj/filesystem.c++:315: failed: expected parts.size() > 0; can't use ".." to break out of starting directory
stack: 7fc5f136d17b 7fc5f136f49a 5583e897a940 7fc5f14fda7d 7fc5f14fdab2 7fc5f14d8eaa 7fc5f14db191 7fc5f14f5571 7fc5f14fd223 7fc5f14fd28e 7fc5f14d9d7a 7fc5f14d99b0 7fc5f14da711 7fc5f14dfffc 7fc5f14e7e8a 7fc5f14e8d02 7fc5f14f9585 7fc5f14fb065 7fc5f14feea6 7fc5f14ff373 5583e8983781 5583e8983ad0 7fc5f1389467 7fc5f138a7a4
make: *** [Makefile:220: capnp/log.capnp.c] Error 1
```

--- a/.changeset/gentle-mayflies-shout.md
+++ b/.changeset/gentle-mayflies-shout.md
@@ -1,0 +1,17 @@
+---
+"miniflare": patch
+---
+
+fix: allow relative `scriptPath`/`modulesRoot`s to break out of current working directory
+
+Previously, Miniflare would resolve relative `scriptPath`s against `moduleRoot` multiple times resulting in incorrect paths and module names. This would lead to `can't use ".." to break out of starting directory` `workerd` errors. This change ensures Miniflare uses `scriptPath` as is, and only resolves it relative to `modulesRoot` when computing module names. Note this bug didn't affect service workers. This allows you to reference a modules `scriptPath` outside the working directory with something like:
+
+```js
+const mf = new Miniflare({
+	modules: true,
+	modulesRoot: "..",
+	scriptPath: "../worker.mjs",
+});
+```
+
+Fixes #4721

--- a/packages/miniflare/src/plugins/core/index.ts
+++ b/packages/miniflare/src/plugins/core/index.ts
@@ -721,8 +721,9 @@ function getWorkerScript(
 	workerIndex: number,
 	additionalModuleNames: string[]
 ): { serviceWorkerScript: string } | { modules: Worker_Module[] } {
-	const modulesRoot =
-		("modulesRoot" in options ? options.modulesRoot : undefined) ?? "";
+	const modulesRoot = path.resolve(
+		("modulesRoot" in options ? options.modulesRoot : undefined) ?? ""
+	);
 	if (Array.isArray(options.modules)) {
 		// If `modules` is a manually defined modules array, use that
 		return {

--- a/packages/miniflare/src/plugins/core/modules.ts
+++ b/packages/miniflare/src/plugins/core/modules.ts
@@ -174,8 +174,6 @@ export class ModuleLocator {
 	}
 
 	visitEntrypoint(code: string, modulePath: string) {
-		modulePath = path.resolve(this.modulesRoot, modulePath);
-
 		// If we've already visited this path, return
 		if (this.#visitedPaths.has(modulePath)) return;
 		this.#visitedPaths.add(modulePath);

--- a/packages/miniflare/test/plugins/core/modules.spec.ts
+++ b/packages/miniflare/test/plugins/core/modules.spec.ts
@@ -1,8 +1,9 @@
 import assert from "assert";
+import fs from "fs/promises";
 import path from "path";
 import test from "ava";
 import { Miniflare, MiniflareCoreError, stripAnsi } from "miniflare";
-import { utf8Encode } from "../../test-shared";
+import { useCwd, useTmp, utf8Encode } from "../../test-shared";
 
 const ROOT = path.resolve(
 	__dirname,
@@ -278,6 +279,30 @@ You must manually define your modules when constructing Miniflare:
     at ${depPath}:2:8`
 	);
 });
+test.serial(
+	"Miniflare: collects modules outside of working directory",
+	async (t) => {
+		// https://github.com/cloudflare/workers-sdk/issues/4721
+		const tmp = await useTmp(t);
+		const child = path.join(tmp, "child");
+		await fs.mkdir(child);
+		await fs.writeFile(
+			path.join(tmp, "worker.mjs"),
+			'export default { fetch() { return new Response("body"); } }'
+		);
+		useCwd(t, child);
+
+		const mf = new Miniflare({
+			modules: true,
+			modulesRoot: "..",
+			scriptPath: "../worker.mjs",
+		});
+		t.teardown(() => mf.dispose());
+
+		const res = await mf.dispatchFetch("http://localhost");
+		t.is(await res.text(), "body");
+	}
+);
 test("Miniflare: suggests bundling on unknown module", async (t) => {
 	// Try with npm-package-like import
 	let mf = new Miniflare({

--- a/packages/miniflare/test/test-shared/storage.ts
+++ b/packages/miniflare/test/test-shared/storage.ts
@@ -32,6 +32,18 @@ export async function useTmp(t: ExecutionContext): Promise<string> {
 	return filePath;
 }
 
+// Must be called from a `.serial()` test
+export function useCwd(t: ExecutionContext, cwd: string) {
+	const originalCwd = process.cwd();
+	const originalPWD = process.env.PWD;
+	process.chdir(cwd);
+	process.env.PWD = cwd;
+	t.teardown(() => {
+		process.chdir(originalCwd);
+		process.env.PWD = originalPWD;
+	});
+}
+
 type ValidReadableStreamBYOBRequest = Omit<
 	ReadableStreamBYOBRequest,
 	"view"

